[NIXL] refactor scheduler->worker request state synchronization

markmc opened the pull request:
In a prefill instance, we need to free KV blocks that have not been fetched after a timeout. See #20139.

In #26012, we're trying to deal with corner cases involved with doing this request timeout tracking on the worker side. This PR proposes refactoring the scheduler->worker request state synchronization to use a SCHEDULED/FINISHED/ABORTED enum rather than in_batch, to_send, and not_processed.

Note the expiry timer is switched back to monotonic time because the timestamp is no longer sent across process boundaries.

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -280,9 +287,14 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
+        # 2. a consumer notification count - with heterogeneous TP, P
+        # must wait for all assigned D TP workers to finish reading
+        # before safely freeing the blocks.
+        self._reqs_need_send: dict[ReqId, tuple[float, int]] = {}
```
I don't love the tuple here

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -395,17 +405,56 @@ def build_connector_meta(
+        for notif in (connector_output.finished_sending or ()):
+            req_id, tp_ratio = notif.decode("utf-8").rsplit(":", 1)
+            # Sent notifications received after we already timed out
+            if req_id not in self._reqs_need_send:
```
Corner case ... we time out (or abort?) and then later receive notifications? Is it possible? We have no test to cover this

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -435,8 +484,15 @@ def request_finished(
+            return False, None
+
+        if request.status != RequestStatus.FINISHED_LENGTH_CAPPED:
+            if request.request_id in self._reqs_need_send:
```
Corner case - a request is aborted after it has been finished? Can this happen? We don't have a unit test that covers it

markmc reviewed (commented):
Some comments on what I'm not fully happy with yet

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -259,6 +266,93 @@ def shutdown(self):
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Invalid notification: {notif}") from e
+
+            # Sent notifications received after we already timed out
```
Corner case ... we time out (or abort?) and then later receive notifications? Is it possible? We have no test to cover this

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -435,8 +529,12 @@ def request_finished(
+        if not params.get("do_remote_decode"):
+            return False, None
+
+        if request.status != RequestStatus.FINISHED_LENGTH_CAPPED:
```
See #25067

markmc replied:
I understand now that this is to handle cases where the abort happens before the request is finished

NickLucche reviewed (commented):
Nice work @markmc !

Personally the major thing that seems a bit off to me is moving the heterogenous-related stuff from Worker->Scheduler. I was hoping we would be able to keep that transparent from the Scheduler, and let the worker handle that as well as the nixl_notifs, which I believe are meant to stay within "nixl agent's reach" (ie Worker-side).
This is also leading to having to keep `consumer_count` in a dataclass, which I believe is making the original changes from last week a bit more complicated.

markmc commented:
> Nice work @markmc !
> 
> Personally the major thing that seems a bit off to me is moving the heterogenous-related stuff from Worker->Scheduler. I was hoping we would be able to keep that transparent from the Scheduler,

It remains hidden from the scheduler itself ... but yes, it's on the scheduler side of the NIXL connector logic

> and let the worker handle that as well as the nixl_notifs, which I believe are meant to stay within "nixl agent's reach" (ie Worker-side).

Nothing jumps out at me as a reason for this being a significant violation of the scheduler/worker split in the connector - the scheduler side is naturally a sort of coordinator, so it being aware that it needs to wait for multiple sent notifications ... doesn't seem breaking any significant encapsulation?

Something needs to keep track of the number of notifications, and in the case of an abort, that counter needs to be freed/deleted ... if the worker is that something, then we can't avoid notifying the worker about the abort

> This is also leading to having to keep `consumer_count` in a dataclass, which I believe is making the original changes from last week a bit more complicated.

For each request, we have a timeout and a consumer count ... I used a tuple initially, but it was a bit gross. It could be two dicts. I dunno ... 

The changes are more significant, because I've refactored and introduced `ReqsNeedSendTracker`, but I'd see the end result as simpler and better encapsulated, rather than more complicated?

markmc committed:
[NIXL] Refactor scheduler->worker request state synchronization

Use a SCHEDULED/FINISHED/ABORTED eNUM rather than in_batch, to_send,
and not_processed.

Also move the expiry timestamp calculation to the worker side
so we're not sending timestamps across process boundaries.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
Looking at moving the consumer notification count stuff back to the worker ... my conclusion is that all of the complexity here comes from having to synchronize the state of requests between the worker and scheduler side, and we will need to do that if the consumer count is tracked by the worker ... so just moving the timeout doesn't achieve anything

So I tried some minor refactoring of the state synchronization

> [NIXL] Refactor scheduler->worker request state synchronization
>    
> Use a SCHEDULED/FINISHED/ABORTED enum rather than in_batch, to_send, and not_processed.
>    
> Also move the expiry timestamp calculation to the worker side so we're not sending timestamps across process boundaries.
